File: backend/src/web/livrocad.py
from __future__ import absolute_import, unicode_literals
from web import my_form
from tekton import router
from livro.model import  Livro
import json


def index(_write_tmpl):

	livros = Livro.query()
	for l in livros:
		l.id = l.key.id()

	var = {"saveLink":"/livrocad/post_salvar",
	       "livros":livros}
	_write_tmpl('templates/admin/livrocad.html', var)

def post_salvar(_handler, autor, titulo, editora, paginas):

	livro = Livro(autor=autor, titulo=titulo, editora=editora, paginas=int(paginas))
	livro.put()

	_handler.redirect("/livrocad")

def salvar(_resp, autor, titulo, editora, paginas):

	livro = Livro(autor=autor, titulo=titulo, editora=editora, paginas=int(paginas))
	key = livro.put() # .put() retorna a key do objeto

	json_str = json.dumps({'id':key.id()})

	_resp.write(json_str)

def listar(_resp):

	json_struct = []

	livros = Livro.query()
	for l in livros:
		json_struct += [{"autor":l.autor,
		        "titulo":l.titulo,
		        "editora":l.editora,
		        "paginas":l.paginas,
		        "idLivro":l.key.id()}]

	json_str = json.dumps(json_struct)
	_resp.write(json_str)

def editar(_resp, idLivro, autor, titulo, editora, paginas):

	livro = Livro.get_by_id(int(idLivro))

	livro.autor = autor
	livro.titulo = titulo
	livro.editora = editora
	livro.paginas = int(paginas)

	livro.put()

	# Não acho que precise de retorno

def remover(_resp, idLivro):

	livro = Livro.get_by_id(int(idLivro))

	livro.key.delete()

File: backend/src/web/static/js/livrocad.js
var myApp = angular.module('myApp', [], function($interpolateProvider) {
	$interpolateProvider.startSymbol('{_');
	$interpolateProvider.endSymbol('_}');
});


function controller($scope, $http){

    $scope.livros = []
 	$http.get('/livrocad/listar').success(function (json){
 		$scope.livros = json;
 	})

  	$scope.editarLivro = function(livro){
  		livro.editando = true;
  	}

 	$scope.salvarLivro = function(){

 		var livro = {"autor":$scope.inputAutor,
 				    "titulo":$scope.inputTitulo,
 				   "editora":$scope.inputEditora,
 				   "paginas":$scope.inputPaginas};

 		$http.post('/livrocad/salvar', livro).success(function (json){

 			livro.idLivro = json.idLivro;
 			livro.editando = false;
 			$scope.livros.push(livro);

 			$scope.inputAutor = "";
 			$scope.inputTitulo = "";
 			$scope.inputEditora = "";
 			$scope.inputPaginas = "";

 		});

 	}

  	$scope.confirmarEdicao = function(livro){
  		livro.editando = false;

 		params =  {"idLivro":livro.idLivro,
 				     "autor":livro.autor,
 				    "titulo":livro.titulo,
 				   "editora":livro.editora,
 				   "paginas":livro.paginas}

 		$http.post('/livrocad/editar', params);
  	}

  	$scope.removerElemento = function(livro, index){
  		$scope.livros.splice(index, 1);
  		livro.editando = false;

 		$http.post('/livrocad/remover', {"idLivro":livro.idLivro});
  	}
}